calyx/src/passes/automatic_par.rs

```diff
@@ -0,0 +1,94 @@
+use crate::lang::component::Component;
+use crate::lang::{ast, ast::Control, context::Context};
+use crate::passes::visitor::{Action, VisResult, Visitor};
+
+/// Pass that collapses
+///(seq
+///    (seq (enable A B)
+///         (enable C D))
+/// ..)
+/// into
+/// (seq (enable A B C D)
+///  ..)
+/// given that there are no edges between the sub-graphs induced by (enable A B) and (enable C D)
+/// since in this case there is no way for these subgraphs to depend on each other
+/// XXX (zhijing): I think this pass need to be changed if we add `enable` CSP style components to futil semantics
+#[derive(Default)]
+pub struct AutomaticPar {}
+
+impl Visitor for AutomaticPar {
+    fn name(&self) -> String {
+        "automatic parallelization".to_string()
+    }
+
+    // use finish_seq so that we collapse things on the way
+    // back up the tree and potentially catch more cases
+    fn finish_seq(
+        &mut self,
+        seq: &mut ast::Seq,
+        comp: &mut Component,
+        _: &Context,
+    ) -> VisResult {
+        let st = &mut comp.structure;
+
+        // get first enable statement and its index. this will act
+        // as the accumulator for the statement that we are collapsing
+        // things into
+        let (start, mut cmp_acc) = match seq.stmts.iter().enumerate().find_map(
+            |(i, stmt)| match stmt {
+                Control::Enable { data } => Some((i, data.clone())),
+                _ => None,
+            },
+        ) {
+            Some((s, c)) => (s, c),
+            None => return Ok(Action::Continue),
+        };
+
+        // vec of new control statements including all elements up to the first
+        // enable that we find
+        let mut new_stmts: Vec<ast::Control> = seq.stmts[..start].to_vec();
+
+        // start interation from the second item because the first
+        // is in `cmp_stmt`
+        for stmt in seq.stmts[start + 1..].iter() {
```

Comment:
Correct me if I understood this code wrong. It is trying to the do the following:
```
(seq
  (enable A)   <-- start
  (enable B)    ---
  (enable C)        | <-- remaining
   . . .        ---
)
```

For every `start` in the sequence, iterate over every other component, detect if they have any components in common. If they don't, parallelize them.

From the loops below, it is hard to parse this intention. Plus, there is a bunch of unnecessary re-computation being done with this pass. When we start checking `B`, we'll have to recompute the components in `C` and so on. I recommend two things:

1. Document the crap out of this pass. Use some [ASCII art](http://asciiflow.com/) and really nail down what this pass should be doing.
2. Consider splitting this into two passes. One that computes all the components on the inputs and outputs of each port, and another one that simply checks this data structure for conflicts. This can also be a good test to see how pass composition works.

Reply:
There's no looping over `start`. `start` just represents the first enable we find. This is just a small optimization to slightly reduce the statements we check.

Doing this in two passes would be less efficient because here we are accumulating enable statements as we go. To show what I mean, suppose
```
(enable A)
(enable B)
(enable C)
```
is equivalent to
```
(enable A B C)
```
In this pass we check if `(enable A)` can be merged with `(enable B)`. In this case they can be so we make the statement `(enable A B)`. Then we check this new statement against `(enable C)` to see if it can be merged. In this case it can, so we produce `(enable A B C)`. etc.

If we did the analysis first, you would have to check all enable statements pairwise, because (using ~ for 'safe to merge with') `(enable A) ~ (enable B) && (enable B) ~ (enable C)` does not imply `(enable A) ~ (enable C)`. To see this imagine 3 components with only an edge between `A` and `C`.